core/src/main/java/io/grpc/internal/AbstractStream2.java

```diff
@@ -0,0 +1,279 @@
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import io.grpc.Codec;
+import io.grpc.Compressor;
+import io.grpc.Decompressor;
+
+import java.io.InputStream;
+
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * The stream and stream state as used by the application. Must only be called from the sending
+ * application thread.
+ */
+public abstract class AbstractStream2 implements Stream {
+  /** The framer to use for sending messages. */
+  protected abstract MessageFramer framer();
+
+  /**
+   * Obtain the transport state corresponding to this stream. Each stream must have its own unique
+   * transport state.
+   */
+  protected abstract TransportState transportState();
+
+  @Override
+  public final void setMessageCompression(boolean enable) {
+    framer().setMessageCompression(enable);
+  }
+
+  @Override
+  public final void writeMessage(InputStream message) {
+    checkNotNull(message);
+    if (!framer().isClosed()) {
+      framer().writePayload(message);
+    }
+  }
+
+  @Override
+  public final void flush() {
+    if (!framer().isClosed()) {
+      framer().flush();
+    }
+  }
+
+  /**
+   * Closes the underlying framer. Should be called when the outgoing stream is gracefully closed
+   * (half closure on client; closure on server).
+   */
+  protected final void endOfMessages() {
+    framer().close();
+  }
+
+  @Override
+  public final void setCompressor(Compressor compressor) {
+    framer().setCompressor(checkNotNull(compressor, "compressor"));
+  }
+
+  @Override
+  public final void setDecompressor(Decompressor decompressor) {
+    transportState().setDecompressor(decompressor);
+  }
+
+  @Override
+  public final boolean isReady() {
+    if (framer().isClosed()) {
+      return false;
```

Comment:
just wanted to say it: I prefer this style over some OR and NEGATE. Much more readable.